src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -671,12 +671,16 @@ def _process_routine_symbols(module_ast, symbol_table,
     '''
     routines = walk(module_ast, (Fortran2003.Subroutine_Subprogram,
                                  Fortran2003.Function_Subprogram))
+    # A subroutine has no type but a function does. However, we don't know what
+    # it is at this stage so we give all functions a DeferredType.
+    type_map = {Fortran2003.Subroutine_Subprogram: NoType(),
+                Fortran2003.Function_Subprogram: DeferredType()}
```

Comment:
That's right. I was imagining it would be a `resolve_deferred` type mechanism that would be needed but I don't yet have a need to do that (although I've remembered #924 which is a use case). I've added a TODO.

Reply:
Created #1314 and updated the TODO.